### PR TITLE
Report all warnings on Werror and fail at the end

### LINKED ISSUE #18634: Not give up on compilation instantly when -Werror is on and warning appears
## Compiler version

3.3.1

## The problem

When warnings are reported in a phase and -Werror (XfatalWarnings) is on, they will be reported as errors and compilation will crash. Because of that, instead of reporting all warnings from linting in a unit, the compilation will exit with an error after just one phase - therefore it does not report warnings from future phases. it would make more sense to aggregate these warnings and report them as errors after the linting phases.

### PR DESCRIPTION
**Pull request by szymon-rd**
Fix #18634 

[test_scala2_library_tasty]

### CODE REVIEW

**Comment by szymon-rd** [comment]
still tbd: massive refactor of tests, but I want to hear if anyone has something against this solution

File: tests/neg-deep-subtype/1828.scala [hunk 1]
```diff
@@ -2,10 +2,12 @@
 
 class Test {
   def remove[S](a: S | Int, f: Int => S):S = a match {
-    case a: S => a             // error
+    case a: S => a             // warn
```
Comment (nicolasstucki) [inline tests/neg-deep-subtype/1828.scala:5]: This tag is not checked by the testing framework. This implies that this file in not testing anything. This implies that all tests that do not have a test file are broken.

We need to make sure that we check the position and number of warnings emitted under `-Xfatal-warnings`/`-Werror`. Our assumption when designing those tests was that we would check them as errors.

**Review by nicolasstucki — DISMISSED** [review]
We need to be able to check if a line emitted a warning or not in the neg tests.

One option would be to make the test framework detect the warnings and `// warn` tags to validate them as we do with `// error`.

Another option would be to make the test framework interpret fatal warnings as errors.

**Comment by nicolasstucki** [comment]
> One option would be to make the test framework detect the warnings and // warn tags to validate them as we do with // error.

The drawback of this approach is that we might need to add more `// warn` in tests. This will be a bit more work on our side but would make tests more robust.

**Comment by dwijnand** [comment]
> One option would be to make the test framework detect the warnings and `// warn` tags to validate them as we do with `// error`.

Move it to `tests/warn/`.

**Comment by nicolasstucki** [comment]
> Move it to `tests/warn/`.

We have tests that check both warnings and errors at the same time.

File: tests/neg-deep-subtype/1828.scala (hunk 1 above)
Comment (szymon-rd) [reply tests/neg-deep-subtype/1828.scala:5]: Dale recently added that feature, it should be working (at least it's in Vulpix in ParallelTesting). I will make sure it works.

Comment (szymon-rd) [reply tests/neg-deep-subtype/1828.scala:5]: Oh. Now I see. It won't test both

**Comment by szymon-rd** [comment]
I can move most of them to warn. For the rest, I will maybe just create the output check files.

**Comment by szymon-rd** [comment]
All the tests are passing now, there are a lot of changes but most are just simple modifications done by scripts.

**Comment by szymon-rd** [comment]
Right now one failing was added, and this branch fails after rebase. But I will fix it after a review, as broken tests keep appearing on main and I will need to fix the ones that appeared before the review is finished anyway.

File: tests/warn/i16649-refutable.scala [hunk 2]
```diff
@@ -1,9 +1,10 @@
...
 
 import quoted.*
 
 def foo(using Quotes)(x: Expr[Int]) =
-  val '{ ($y: Int) + ($z: Int) } = x // error
+  val '{ ($y: Int) + ($z: Int) } = x // warn
   val '{ $a: Int } = x
   val '{ $b: Any } = x
   val '{ $c } = x
+
```
Comment (nicolasstucki) [inline tests/warn/i16649-refutable.scala:10]:
```suggestion
```

**Review by nicolasstucki — COMMENTED** [review]
We can simplify the work needed to get this PR considerably if we split it in parts.

1. Add warn test infrastructure
2. Migrate neg tests into warn tests (in batches)
3. Change warning/error behavior

**Comment by szymon-rd** [comment]
Good idea. I will split it in a couple of PRs.

**Pull request closed without merging**